### DISCUSSION TITLE
All things React

### DISCUSSION THREAD

**Discussion by drteresavasquez** [nss-evening-cohort-14/evening-client-side#218] [Class Notes and Discussion]
# All about that React!

This ticket is reserved for all things React that we will be doing in class. Here is the video page for all the videos we have done in class. Also, the password to view the videos will be shared with you separately.

## Links
- [Videos](https://vimeo.com/showcase/reactwithdrt)
- [React Template](https://github.com/nss-evening-cohort-14/react-template)
- [Reactstrap](https://reactstrap.github.io/)
- [React Semantic UI](https://react.semantic-ui.com/)
- [Intro to React Slides](https://docs.google.com/presentation/d/1mDCk7-dqwbnmDN-K2zgpwINeTHtKatBdhCjzxELmvXU/edit?usp=sharing)

## May 15, 2021: Styled Components
- Check out this repo for starter: https://github.com/drteresavasquez/styled-components
- Styled Components Docs: https://styled-components.com/docs/basics#installation
- Sample Todo App using Styled Components: https://react-todo-this.netlify.app/
- Github repo: https://github.com/nss-evening-cohort-14/react-less-do-dis
- Flow Chart of Todo App: https://docs.google.com/presentation/d/1SD7nHibdA5H63wVc3NZtcQRA1cgqUNc8awxlkj0-pq4/edit?usp=sharing
<img width="1134" alt="Screen Shot 2021-05-16 at 11 14 39 AM" src="https://user-images.githubusercontent.com/29741570/118404295-fb452000-b637-11eb-8890-56d3f4c66424.png">

## May 3, 2021: Adding Auth & Private Routes
- [Pull request for adding auth](https://github.com/nss-evening-cohort-14/all-things-react/pull/4/files)
### Private Routes Broiler Plate Component:

```javascript
// The PrivateRoute function is creating a private route and returing the specified route based on the props

// We specify the specific props we want to use in the routeChecker and pass the rest with the spread
const PrivateRoute = ({ component: Component, user, ...rest }) => {
  // when we call this function in the return, it is looking for an argument. `props` here is taco.
  const routeChecker = (taco) => (user
    ? (<Component {...taco} user={user} />)
    : (<Redirect to={{ pathname: '/', state: { from: taco.location } }} />));
    // this render method is one we can use instead of component. Since the components are being dynamically created, we use render. Read the docs for more info: https://reactrouter.com/web/api/Route/render-func

  // Just like in the routes if we want the dynamically rendered component to have access to the Router props, we have to pass `props` as an argument.
  return <Route {...rest} render={(props) => routeChecker(props)} />;
};
```

## May 1, 2021: React Router Dom - Dynamic Routes
<img width="1241" alt="Screen Shot 2021-05-01 at 1 47 48 PM" src="https://user-images.githubusercontent.com/29741570/116792126-5dbcee80-aa84-11eb-93a0-c23d4b6de6a3.png">

## April 27, 2021: React Router DOM
https://reactrouter.com/web/guides/quick-start

[React router cheat sheet](https://drive.google.com/file/d/1ySy7igaShG_BtXHFL-oPxTeckyumiARj/view?usp=drivesdk)

Use the React forms repo we have been working out of.
React Router DOM PR: https://github.com/nss-evening-cohort-14/all-things-react/pull/3

### To get started
- run `npm i react-router-dom`
- Create `src/views` directory 
- In views, create 3 files:
  - `Home.js` - create react functional component that returns an `h1`
  - `Students.js`
  - `AddStudent.js` - create react functional component that returns an `h1`
- Copy/paste `src/App/index.js` code to `src/views/Students.js`
- In `src/App/index.js` remove everything except the return statement. Have the file return `<h1>Hello World</h1>` 
<details>
  <summary>Click to view code for `src/Apps/index.js`</summary>

```javascript
import React from 'react';

function App() {
  return (
    <h1>Hello World!</h1>
  );
}

export default App;
```
</details>

- Create a new component: `src/components/NavBar.js`

<details>
  <summary>Click to view code for `src/components/NavBar.js`</summary>
  
```javascript
import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
} from 'reactstrap';

const NavBar = () => {
  const [isOpen, setIsOpen] = useState(false);

  const toggle = () => setIsOpen(!isOpen);

  return (
    <div>
      <Navbar color="light" light expand="md">
        <NavbarBrand href="/">React</NavbarBrand>
        <NavbarToggler onClick={toggle} />
        <Collapse isOpen={isOpen} navbar>
          <Nav className="mr-auto" navbar>
            <NavItem>
              <Link className="nav-link" to="/add-student">Add Student</Link>
            </NavItem>
            <NavItem>
              <Link className="nav-link" to="/students">Student Cards</Link>
            </NavItem>
          </Nav>
        </Collapse>
      </Navbar>
    </div>
  );
};

export default NavBar;

```
</details>

### App should like this when done:
<img width="591" alt="Screen Shot 2021-04-27 at 3 56 01 PM" src="https://user-images.githubusercontent.com/29741570/116313725-bd10bb00-a773-11eb-91de-47f2ac05f297.png">

___
## April 25, 2021: Passing props to update the DOM, Delete, and Update
Link to code: https://github.com/nss-evening-cohort-14/all-things-react

### In Class Assignment
1. Add the `Delete` button to authors and do all the things
2. Add the DOM reload on form submit of new student
3. Add the Edit button to the author and implement the form and all the things... 

Flow chart of props
<img width="1321" alt="Screen Shot 2021-04-27 at 9 20 51 AM" src="https://user-images.githubusercontent.com/29741570/116257731-e9a7e100-a739-11eb-9e39-a38b5b2f8008.png">

Some tricks for passing props: https://github.com/nss-evening-cohort-14/all-things-react/pull/2

___
## April 24, 2021: Forms and PropTypes
By the end of this class, you will understand how to create forms, pass data to functions, render to the DOM from data created in the form, how to apply proptypes to help you with testing your components and how to delete. 

Using the React Template, complete the following:

1. Create a new repo called react-forms
2. Create a new repo called react-almost-amazon
3. Create a new Firebase project called react-forms so that you can post your new students
4. Copy/paste all your API keys in the .env file

### In Class Assignment
1. Forms: create a new form and make all the necessary connections in the new react-almost-amazon repo. Use your OLD almost Amazon Firebase project. 
2. PropTypes: create a author component card using proptypes
3. Passing props: pass author array after adding a new author back up to app. 

Link to code: https://github.com/nss-evening-cohort-14/all-things-react

___
## April 20, 2021: API Lab
By the end of class tonight, you should at least have the Joke Generator complete. If you get that one done, move on to the next app on the list: https://github.com/nss-evening-cohort-14/evening-client-side/discussions/204

___
## April 19, 2021: Airbender API Quiz App
Tonight we worked on `useEffect` and more with `useState` hooks. 
Link to Repo: https://github.com/drteresavasquez/airbender-quiz

___
## April 17, 2021: Counter App
Created our first React App - A counter
Repo: https://github.com/drteresavasquez/intro-to-react-2

**Comment by jrobinson0529** [comment]
Found an easy and pretty NPM package for animating components on react! Very simple to use.
https://nearform.github.io/react-animation/

**Comment by drteresavasquez** [comment]
Cool select menu package: https://react-select.com/home
